conversion/resource_generation.go

```diff
@@ -0,0 +1,401 @@
+	}
+
+	return common.TaskResult[*ConnectionProfileReq]{Result: createResourceData, Err: nil}
+}
+
+// If any of the resource creation fails, deletes all resources that were created
+func (r ResourceGenerationImpl) ConnectionProfileCleanUp(ctx context.Context, profiles []*ConnectionProfileReq) error {
+	for _, profile := range profiles {
+		err := r.DsAcc.DeleteConnectionProfile(ctx, r.DsClient, profile.ConnectionProfile.ProjectId, profile.ConnectionProfile.Region, profile.ConnectionProfile.Id)
+		if err != nil {
+			return err
+		}
+
+		if profile.ConnectionProfile.BucketName != "" {
+			err := r.StorageAcc.DeleteGCSBucket(ctx, r.StorageClient, storageaccessor.StorageBucketMetadata{BucketName: profile.ConnectionProfile.BucketName})
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Returns source and destination connection profiles to be created
+func (r ResourceGenerationImpl) GetResourcesForCreation(ctx context.Context, projectId string, sourceProfile profiles.SourceProfile, region string, validateOnly bool) ([]*ConnectionProfileReq, []*ConnectionProfileReq, error) {
+	var sourceProfilesToCreate []*ConnectionProfileReq
+	var dstProfilesToCreate []*ConnectionProfileReq
+
+	// Map for each region with list of all connection profiles
+	var connectionProfiles map[string][]string = make(map[string][]string)
+	var err error = nil
+
+	for _, profile := range sourceProfile.Config.ShardConfigurationDataflow.DataShards {
+		// Check if source profile needs to be created
+		sourceProfile, err := getSourceConnectionProfileForCreation(ctx, projectId, profile, region, validateOnly, connectionProfiles, r.DsAcc, r.DsClient)
+		if err != nil {
+			return sourceProfilesToCreate, dstProfilesToCreate, err
+		}
+		if sourceProfile != nil {
+			sourceProfilesToCreate = append(sourceProfilesToCreate, sourceProfile)
+		}
+
+		// Check if destination profile needs to be created
+		dstProfile, err := getDstConnectionProfileForCreation(ctx, projectId, profile, region, validateOnly, connectionProfiles, r.DsAcc, r.DsClient)
+		if err != nil {
+			return sourceProfilesToCreate, dstProfilesToCreate, err
+		}
+		if dstProfile != nil {
+			dstProfilesToCreate = append(dstProfilesToCreate, dstProfile)
+		}
+	}
+	return sourceProfilesToCreate, dstProfilesToCreate, err
+}
+
+// 1. For each datashard, check if source and destination connection profile exists or not
+// 2. If source connection profile doesn't exists create it or validate if creation is possible.
+// 3. If validation is false and destination connection profile doesn't exists create a corresponding gcs bucket and then a destination connection profile
+func (c *ValidateOrCreateResourcesImpl) ValidateOrCreateResourcesForShardedMigration(ctx context.Context, projectId string, instanceName string, validateOnly bool, region string, sourceProfile profiles.SourceProfile) error {
```

Comment:
Both this method and the interface is doing a lot. Is there a challenge to break this into `validate` and `create` as different methods?